rclcpp/include/rclcpp/qos_event.hpp

```diff
@@ -0,0 +1,164 @@
+{
+
+struct QOSDeadlineEventInfo
+{
+  enum DeadlineEventType
+  {
+    DEADLINE_MISSED
+  };
+
+  DeadlineEventType event_type;
+  void * other_info;
+};
+
+struct QOSLivelinessEventInfo
+{
+  enum LivelinessEventType
+  {
+    LIVELINESS_CHANGED
+  };
+
+  LivelinessEventType event_type;
+  void * other_info;
+};
+
+struct QOSLifespanEventInfo
+{
+  enum LifespanEventType
+  {
+    LIFESPAN_EXPIRED
+  };
+
+  LifespanEventType event_type;
+  void * other_info;
+};
+
+
+using QOSDeadlineEventCallbackType = std::function<void (QOSDeadlineEventInfo &)>;
+using QOSLivelinessEventCallbackType = std::function<void (QOSLivelinessEventInfo &)>;
+using QOSLifespanEventCallbackType = std::function<void (QOSLifespanEventInfo &)>;
+
+
+// struct QOSDeadlineEventConfig {
+//   using CallbackT = QOSDeadlineEventCallbackType;
+//   using EventInfoT = QOSDeadlineEventInfo;
+// };
+
+// struct QOSLivelinessEventConfig {
+//   using CallbackT = QOSLivelinessEventCallbackType;
+//   using EventInfoT = QOSLivelinessEventInfo;
+// };
+
+// struct QOSLifespanEventConfig {
+//   using CallbackT = QOSLifespanEventCallbackType;
+//   using EventInfoT = QOSLifespanEventInfo;
+// };
+
+
+class QOSEventBase : public Waitable
+{
+public:
+  /// Get the number of ready events
+  size_t
+  get_number_of_ready_events() override;
+
+  /// Add the Waitable to a wait set.
+  bool
+  add_to_wait_set(rcl_wait_set_t * wait_set) override;
+
+  /// Check if the Waitable is ready.
+  bool
+  is_ready(rcl_wait_set_t * wait_set) override;
+
+protected:
+  rcl_event_t event_handle_;
+  size_t wait_set_event_index_;
+};
+
+
+template<typename EventCallbackT>
+class QOSEvent : public QOSEventBase
+{
+public:
+  template<typename InitFuncT, typename HandleT, typename OptionsT, typename EventTypeEnum>
+  QOSEvent(
+    const EventCallbackT & callback,
+    InitFuncT init_func,
+    HandleT handle,
+    OptionsT options,
+    EventTypeEnum type)
+  : event_callback_(callback)
+  {
+    event_handle_ = rcl_get_zero_initialized_event();
+    rcl_ret_t ret = init_func(&event_handle_, handle, options, type);
+    if (ret != RCL_RET_OK) {
+      rclcpp::exceptions::throw_from_rcl_error(ret, "could not create event");
+    }
+  }
+
+  ~QOSEvent()
+  {
+    if (rcl_event_fini(&event_handle_) != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rclcpp",
+        "Error in destruction of rcl event handle: %s", rcl_get_error_string().str);
+      rcl_reset_error();
+    }
+  }
+
+  /// Execute any entities of the Waitable that are ready.
+  void
+  execute() override
+  {
+    EventCallbackInfoT callback_info;
+
+    rcl_ret_t ret = rcl_take_event(&event_handle_, callback_info.other_info);
+    if (ret != RCL_RET_OK) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rclcpp",
+        "Couldn't take event info: %s", rcl_get_error_string().str);
+      return;
+    }
+
+    event_callback_(callback_info);
+  }
+
+private:
+  using EventCallbackInfoT = typename std::remove_reference<typename
```

Comment:
That's a dense bit of templating magic to parse. :)

Reply:
This template magic is provided in `#include "rclcpp/function_traits.hpp"`, so it must be sanctioned :)